### DISCUSSION TITLE
How much CO2/GHG does Copilot emit?

### DISCUSSION THREAD

**Discussion by merlinaudio** [community/community#38168] [Copilot Conversations]
I'm wondering how much energy the use of GitHub Copilot costs and how much CO2/GHG it emits.

I realize GitHub is carbon neutral, but what about Copilot? If GitHub wasn't carbon neutral, how much would we be looking at here?

Getting some stats about the training would be great, too, as I've read articles about GPT-3 training emitting [**3 times the "emissions of a round trip flight of a passenger jet between San Francisco and New York"**](https://www.cdotrends.com/story/15741/ai-training-scale-might-be-more-efficient-we-think)

Are there projections for how this might change in the coming years?

Much appreciated!

**Comment by david-engelmann** [comment]
How much energy is used compiling broken code?

**Reply by merlinaudio** [reply]
Depends -- for languages like JavaScript, Go, etc I wouldn't think it's that much.

I don't even want to know the number of hours spent compiling broken C++, though :-P

**Reply by JonasDoe** [reply]
I doubt that running a compilation competes with triggering a generative transformer on almost every key-stroke. It's not just its (regular) training - which requires immense resources -, but also the daily drain for all these computed trivialities on our machines.

**Reply by david-engelmann** [reply]
why do you doubt this, its not like we train a model on each key stroke? The majority of my projects/application use significantly more resources to compile then they do to make 1 query against an LLM

**Reply by JonasDoe** [reply]
No, for all I know, they train the model, and afterward we use it. You can't provide a result and learn from it at the same time - you'ld feed the output into the input, i.e. train the model with its own result. Actually, it's a huge issue for all model training that the internet is rather polluted with AI content by now, which must be used for further training.

**Reply by david-engelmann** [reply]
You aren't training the model with the previous response, you are prompting the model with the previous results. Whether your input ends up training a model is dependent on your LLM provider and your agreements with them. Using an LLM for an autocomplete or a code suggestions has a resource usage equivalent to a google search (without considering the resources used to run Chrome on your machine).

**Reply by JonasDoe** [reply]
Gotta do more research on that. But note that my main argument is that it's triggered by many keystrokes, for the whole working day.

**Reply by david-engelmann** [reply]
> Gotta do more research on that. But note that my main argument is that it's triggered by many keystrokes, for the whole working day.

most likely you are running tree-sitter and a lsp that are doing a ton of things on every keystroke, even suggest and autocompleting, the additional resources used to hit an LLM are trival compared to failing to compile on each keystroke until you type out the snippet you could have generated. autocompleting the code you need is using less energy then spending 10 minutes writing that code, specially if you are googling, clicking through docs, letting things idle is energy as well.

**Reply by JonasDoe** [reply]
Alright, a couple of things:
- Regarding the training, we were talking about different things. Both apply, ofc.: the pretraining of the model, as well as the "personalization" training done by your LLM provider
- The power consumption on use differs probably quite a lot, depending which LLM is used. What I've heart of GPT (and I figure Microsoft is utilizing that by now), it's rather intense on calculations b/c it needs to run over and over again for each generated word/token. If you have more knowledge here than I have, I'ld happy to hear about it, though (no sarcasm intended)!
- You have a good point here that keystrokes in IDEs are probably enforcing rather much calculations anyway. I'm not sure agree that the use of Copilot is a balanced trade-off here. I fully agree that if you start with googling, reading, writing for new, non-straight forward functions, generating it with Copilot might be advantagous. But I'm having my doubts about a simple autocompletion feature for words, lines or simple clauses. These won't take 10 minutes to write and neither require much googling, but a couple of seconds, and your IDE just have to lookup for keywords and (often) just the autocompletion derived from static types. Our job profiles might differ here, but speaking for me, most of the daily work consists of the latter - I rarely have to come up with a new algorithm worth the name.

Lots of speculation here, therefore hearing some numbers from Microsoft/Github would be nice. I've read that Microsoft is heading for a solution base on nuclear energy to satisfeed the ever-increasing need stemming from AI ([source](https://www.windowscentral.com/microsoft/microsoft-is-reportedly-eyeing-nuclear-energy-for-its-ai-ventures-following-the-techs-exorbitant-power-consumption)), but ofc. you can't account all of that to the use of Copilot, and as you pointed out: we can't just ignore that power is spent at other places if we dropped Copilot, so it's not really insightful on its own.

**Reply by david-engelmann** [reply]
Training wise you are hundred percent correct, pretraining a model is energy expended once for all the downstream users of the model, the energy spent to create chatgpt-3.5 is spread out by every user of that model for each call (per request a very small number). The "personalization" is based on the LLM provider and your contract with them. If you've paid for a premium service such as Sourcegraph they will create a graph representation of your code (think tree-sitter). When its time for the LLM to generate an answer, they take the question asked and do a RAG search on the graph representation of the code and they prompt the generally trained model with the relevant code its retrieved from the RAG search and your question (or in the case of autocomplete, what's been typed already). While RAG search is vector math and not without a resource cost, modern machine are really good at these calculations. The frequency that call is made is going to be based on the tool you use, whether that's per key stroke, per word or whatnot (Sourcegraph for example can be a chat triggered calls from within the IDE, which only occur when you interact with the chat). There may be tools that are creating custom trained LLM models based on your code base but they aren't the typical solution, as that process is expensive and the improvement to the results would be minimal and become dated very quickly. Comparing a developers resource usage with and without LLM-based tools are difficult and is widely influenced by the type of programmer. Someone using lots of third-party tools with lots of functions they've not wrote could see huge benefits in a tool that immediately tells you what function to call and what arguments to pass and what those arguments correspond to in the current scope. If you already wrote the code many times. then yes copy and pasting it will be more efficient. That's not where my time is spent. All this being said it makes the assumption that the LLM generated code is valid and efficient or the programmer using the tools has the ability to notice and correct it's deficiency easily. For experienced users, its probably more efficient, for novice its probably a huge time and resource suck

**Reply by JonasDoe** [reply]
Oh, so understand correctly that Copilot's proposals don't require a full-fledged GPT request under all circumstances, but often just rely on a Graph-based search powered by LLM provider? I tried to get an idea what RAG search is doing: looks like it uses GPT to transform its result into an adequate, request-related response. But doesn't this last step mean the same power consumption as if we queried GPT directly, plus the - probably neglectable - power required for the Graph search on top?

**Reply by david-engelmann** [reply]
The power consumption is the graph search to get the relevant code for the query and the LLM query itself. It depends on how the plugin is setup. my assistant makes a query when i stop typing for a quarter of a second, while typing the previous queries results are present in the drop down until I pause typing and it refreshes the results.

**Reply by JonasDoe** [reply]
I see, but this is what I'm expecting to be rather heavy on power consumption - lots of queries against the LLM.

**Comment by cmp-nct** [comment]
Given that CO2 is the second most important food of plants, right after sunlight I do hope my use of Copilot does emit a lot of extra CO2.
I like plants, I also like food. It's win-win.

Though I fear the impact is neglectable :( After all training is only done rarely and then millions of users use the finished models.
Also a single jet round trip sounds really like not enough :( After all aviation produces just 2% of the annual human CO2 (all trips together).

We will need to look at another solution for generating plant food! (Do you run a greenhouse by chance ?)

**Reply by zuckerruebe** [reply]
Your plants don't like it so much when the temperature changes too fast for them to evolve with it. Neither do I.

**Reply by ptomaudio** [reply]
That's the problem though, we as a species have destroyed so many plants in our history that there's far more food than there are plants to eat it. My wife likes a chocolate brownie but if I bought home a box of 100 brownies every day and we didnt get rid of them by other means, then we'd have issues after a while

**Comment by gokhan-or** [comment]
> I'm wondering how much energy the use of GitHub Copilot costs and how much CO2/GHG it emits.
> 
> I realize GitHub is carbon neutral, but what about Copilot? If GitHub wasn't carbon neutral, how much would we be looking at here?
> 
> Getting some stats about the training would be great, too, as I've read articles about GPT-3 training emitting [**3 times the "emissions of a round trip flight of a passenger jet between San Francisco and New York"**](https://www.cdotrends.com/story/15741/ai-training-scale-might-be-more-efficient-we-think)
> 
> Are there projections for how this might change in the coming years?
> 
> Much appreciated!

**Comment by ahengst** [comment]
Food for thought: (not an answer) Your Q feeds into a bigger question: could IP or HTTP be modified to reflect the energy used* to produce a response?  I've thought/discussed this enough to know it's probably impossible to pull off. But at least one RFC was proposed - it failed (I think it was flawed). 
*code and CPUs are able to estimate energy much more readily than CO2 byproducts, which vary depending on grid. KISS principle.
*reducing energy (even green energy) consumption will indirectly reduce fossil energy production.

We live in a world of energy blindness and I've been putting off using ANY kind of AI because it just sounds terrible from an energy perspective. It's OK to ask about working "harder not smarter", but we can't do that without having any solid numbers either way.

**Reply by ahengst** [reply]
-- I'm writing this just as I'm (finally) signing up for copilot. Useful comments above, thx.
-- Whether github is "CO2 neutral" does not change the fact that emissions are occurring and could be reduced.

**Comment by ycouble** [comment]
I do not have the complete answer, but here are some sourced elements that can help thinking about this problem :
- Yes training an LLM is energetically intensive and produces more or less carbon depending on the grid CO2 intensity of the location where is is run. GPT-3 and BLOOM, two LLMs that are probably similar in size to GH Copilot (?), required respectively 1287 MWh and 433 MWh, generating respectively 522 tCO2eq (US) and 30 tCO2eq (France) ([source](https://arxiv.org/pdf/2211.02001) and [gpt-3 source](https://arxiv.org/pdf/2104.10350))
- Inference is **not** negligible and neither equivalent to a simple API request. LLM inference requires to run full forward passes through the neural network which represent a huge amount of computation, and running an LLM for inferences is still highly hardware intensive (BLOOM required running on 16 GPU (A100 80GB) for inference). ([same source](https://arxiv.org/pdf/2211.02001))
- more recent and larger LLMs are much more hardware and energy intensive (GPT-4 expected to be running on 128GPUs clusters for inference)
- Estimates about the energy consumption of a LLM query are still quite rare, but i calculated from [the bloom experiment paper data](https://arxiv.org/pdf/2211.02001) an **energy cost of 3.96 Wh / request** and corresponding **carbon emissions 1,56 gCO2eq / request** (GCP US-central), to compare to the 0.3Wh average energy used for an average google request [reported by google here in an NYT article](https://www.nytimes.com/2011/09/09/technology/google-details-and-defends-its-use-of-electricity.html) which is far from a simple API request.

**Reply by JonasDoe** [reply]
Thanks for these thoughts! Sounds rather discouraging. :see_no_evil:

**Reply by ptomaudio** [reply]
It's crazy how long it takes for discussions like these to result in someone actually providing researched evidence instead of arguing based on 'trust me bro'
Bravo

**Comment by ycouble** [comment]
There’s a community driven estimator for environnemental impact being built right now: ecologits.ai, they try to estimate co2eq and kwh in production scenarios, on a per-token basis.

They are also starting to dig into code completion tools, although i’m not sure how far they went on the topic.
